src/main/java/org/eclipse/yasson/internal/serializer/types/EnumSerializer.java

```diff
@@ -12,26 +12,57 @@
 
 package org.eclipse.yasson.internal.serializer.types;
 
+import java.util.EnumMap;
+
 import jakarta.json.stream.JsonGenerator;
 
 import org.eclipse.yasson.internal.SerializationContextImpl;
+import org.eclipse.yasson.internal.model.ClassModel;
+import org.eclipse.yasson.internal.model.PropertyModel;
+
 
 /**
  * Serializer of the {@link Enum} types.
  */
 class EnumSerializer extends TypeSerializer<Enum<?>> {
 
+    private final EnumMap<? extends Enum<?>, String> constantToNameMap;
+
     EnumSerializer(TypeSerializerBuilder serializerBuilder) {
         super(serializerBuilder);
+
+        constantToNameMap = createConstantToNameMap(serializerBuilder);
+    }
+
+    private static <E extends Enum<E>> EnumMap<E, String> createConstantToNameMap(TypeSerializerBuilder serializerBuilder) {
+        EnumMap<E, String> constantToNameMap = null;
+        Class<?> clazz = serializerBuilder.getClazz();
+
+        if (clazz.isEnum()) {
+            try {
+                @SuppressWarnings("unchecked")
+                Class<E> enumClazz = (Class<E>) clazz;
+                constantToNameMap = new EnumMap<>(enumClazz);
+                ClassModel classModel = serializerBuilder.getJsonbContext().getMappingContext().getOrCreateClassModel(clazz);
+
+                for (E enumConstant : enumClazz.getEnumConstants()) {
+                    PropertyModel model = classModel.getPropertyModel(enumConstant.name());
+                    constantToNameMap.put(enumConstant, model.getWriteName());
+                }
+            } catch (ClassCastException classCastException) {
+                throw new IllegalArgumentException("EnumSerializer can only be used with Enum types");
```

Comment:
This should never happen, if method isEnum is used.

Reply:
The CCE could happen here not because I cast the class to the enum class, but to some definite enum class E in line 43. Otherwise, I wouldn't have typed constants in the loop. If you think that this wouldn't happen or cause the exception - I can remove the try-catch block. Or if you think that it's OK to have here a CCE been thrown and not IAE.